File: src/com/intellij/idea/plugin/hybris/toolwindow/RemoteHacConnectionDialog.java
package com.intellij.idea.plugin.hybris.toolwindow;

import com.intellij.idea.plugin.hybris.common.services.CommonIdeaService;
import com.intellij.idea.plugin.hybris.common.utils.HybrisI18NBundleUtils;
import com.intellij.idea.plugin.hybris.notifications.Notifications;
import com.intellij.idea.plugin.hybris.settings.HybrisRemoteConnectionSettings;
import com.intellij.idea.plugin.hybris.tools.remote.http.HybrisHacHttpClient;
import com.intellij.idea.plugin.hybris.toolwindow.document.filter.UnsignedIntegerDocumentFilter;
import com.intellij.idea.plugin.hybris.toolwindow.document.listener.SimpleDocumentListener;
import com.intellij.notification.NotificationType;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.ui.components.OnOffButton;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.event.DocumentEvent;
import javax.swing.text.PlainDocument;
import java.awt.*;
import java.util.Objects;

import static com.intellij.openapi.ui.DialogWrapper.IdeModalityType.PROJECT;

public class RemoteHacConnectionDialog extends DialogWrapper {

    private JPanel contentPane;
    private JTextField projectIpTextField;
    private JLabel projectIpLabel;
    private JTextField loginTextField;
    private JPasswordField passwordField;
    private JButton testConnectionButton;
    private JLabel loginNameLabel;
    private JLabel passwordLabel;
    private JLabel projectPortLabel;
    private JTextField projectPortTextField;
    private JLabel projectUrlPreviewValueLabel;
    private JLabel projectUrlPreviewLabel;
    private JLabel hacWebrootLabel;
    private JTextField hacWebrootTextField;
    private JTextField displayNameTextField;
    private OnOffButton sslButton;
    private JLabel sslLabel;
    private JComboBox sslProtocol;
    private JLabel sslProtocolLabel;

    private Project myProject;
    private HybrisRemoteConnectionSettings mySettings;
    private HybrisRemoteConnectionSettings setting;

    public RemoteHacConnectionDialog(
        @Nullable final Project project,
        @Nullable final Component parentComponent,
        @NotNull final HybrisRemoteConnectionSettings settings
    ) {
        super(project, parentComponent, false, PROJECT);
        myProject = project;
        mySettings = settings;

        displayNameTextField.setText(mySettings.getDisplayName());
        projectIpTextField.setText(mySettings.getHostIP());
        projectPortTextField.setText(mySettings.getPort());
        ((PlainDocument) projectPortTextField.getDocument()).setDocumentFilter(new UnsignedIntegerDocumentFilter());
        hacWebrootTextField.setText(mySettings.getHacWebroot());
        loginTextField.setText(mySettings.getHacLogin());
        passwordField.setText(mySettings.getHacPassword());
        sslButton.setSelected(mySettings.isSsl());
        if (mySettings.getSslProtocol() != null) {
            sslProtocol.setSelectedItem(mySettings.getSslProtocol());
        }
        setting = new HybrisRemoteConnectionSettings();
        saveSettings(settings);

        final SimpleDocumentListener validateDocumentListener = new SimpleDocumentListener() {

            @Override
            public void update(final DocumentEvent e) {
                validateParams();
            }
        };

        final SimpleDocumentListener saveSettingsDocumentListener = new SimpleDocumentListener() {
            @Override
            public void update(final DocumentEvent e) {
                saveSettings(setting);
            }
        };

        validateParams();
        init();

        projectPortTextField.getDocument().addDocumentListener(validateDocumentListener);
        projectIpTextField.getDocument().addDocumentListener(validateDocumentListener);
        displayNameTextField.getDocument().addDocumentListener(saveSettingsDocumentListener);
        projectIpTextField.getDocument().addDocumentListener(saveSettingsDocumentListener);
        projectPortTextField.getDocument().addDocumentListener(saveSettingsDocumentListener);
        hacWebrootTextField.getDocument().addDocumentListener(saveSettingsDocumentListener);
        displayNameTextField.addActionListener(action->saveSettings(setting));
        projectIpTextField.addActionListener(action->saveSettings(setting));
        hacWebrootTextField.addActionListener(action->saveSettings(setting));
        loginTextField.addActionListener(action->saveSettings(setting));
        passwordField.addActionListener(action->saveSettings(setting));
        sslButton.addActionListener(action-> toggleSslButton(setting));
        sslProtocol.addActionListener(action->saveSettings(setting));

        Objects.requireNonNull(getButton(getOKAction())).addActionListener(action->saveSettings(mySettings));
        testConnectionButton.addActionListener(action -> testConnection());
    }

    private void testConnection() {
        saveSettings(setting);
        HybrisHacHttpClient hybrisHacHttpClient = HybrisHacHttpClient.getInstance(myProject);
        final String errorMessage = hybrisHacHttpClient.login(myProject, setting);
        final String testedHacURL = hybrisHacHttpClient.getHostHacURL(myProject, setting);

        final NotificationType type;
        final String message;
        if (errorMessage.isEmpty()) {
            message = HybrisI18NBundleUtils.message("hybris.toolwindow.hac.test.connection.success", "hac", testedHacURL);
            type = NotificationType.INFORMATION;
        } else {
            type = NotificationType.WARNING;
            message = HybrisI18NBundleUtils.message("hybris.toolwindow.hac.test.connection.fail", testedHacURL, errorMessage);
        }

        Notifications.create(type, HybrisI18NBundleUtils.message("hybris.notification.toolwindow.hac.test.connection.title"), message)
                     .notify(myProject);
    }

    private void toggleSslButton(final HybrisRemoteConnectionSettings setting) {
        sslProtocol.setEnabled(sslButton.isSelected());
        saveSettings(setting);
    }

    private void saveSettings(final HybrisRemoteConnectionSettings mySettings) {
        mySettings.setType(HybrisRemoteConnectionSettings.Type.Hybris);
        mySettings.setSsl(sslButton.isSelected());
        mySettings.setSslProtocol((String) sslProtocol.getSelectedItem());
        mySettings.setDisplayName(displayNameTextField.getText());
        mySettings.setHostIP(projectIpTextField.getText());
        mySettings.setPort(projectPortTextField.getText());
        mySettings.setHacWebroot(hacWebrootTextField.getText());
        mySettings.setHacLogin(loginTextField.getText());
        mySettings.setHacPassword(new String(passwordField.getPassword()));
        final String previewUrl = CommonIdeaService.Companion.getInstance().getHostHacUrl(myProject, mySettings);
        projectUrlPreviewValueLabel.setText(previewUrl);
        mySettings.setGeneratedURL(previewUrl);
    }

    @Nullable
    @Override
    protected JComponent createCenterPanel() {
        return contentPane;
    }

    private void validateParams() {
        testConnectionButton
                .setEnabled(!projectPortTextField.getText().isEmpty()
                        && !projectIpTextField.getText().isEmpty());

        getOKAction().setEnabled(!projectPortTextField.getText().isEmpty()
                && !projectIpTextField.getText().isEmpty());
    }

    private void createUIComponents() {
    }
}

File: src/com/intellij/idea/plugin/hybris/toolwindow/RemoteHacConnectionDialog.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for IntelliJ IDEA.
 * Copyright (C) 2019-2024 EPAM Systems <hybrisideaplugin@epam.com> and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.toolwindow

import com.intellij.idea.plugin.hybris.common.HybrisConstants
import com.intellij.idea.plugin.hybris.common.utils.HybrisI18NBundleUtils.message
import com.intellij.idea.plugin.hybris.notifications.Notifications
import com.intellij.idea.plugin.hybris.settings.HybrisRemoteConnectionSettings
import com.intellij.idea.plugin.hybris.tools.remote.RemoteConnectionScope
import com.intellij.idea.plugin.hybris.tools.remote.RemoteConnectionUtil
import com.intellij.idea.plugin.hybris.tools.remote.http.HybrisHacHttpClient
import com.intellij.notification.NotificationType
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.ComboBox
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.ui.EnumComboBoxModel
import com.intellij.ui.SimpleListCellRenderer
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBPasswordField
import com.intellij.ui.components.JBTextField
import com.intellij.ui.dsl.builder.*
import com.intellij.ui.layout.selected
import com.intellij.util.ui.JBUI
import java.awt.Component
import java.awt.event.ActionEvent
import java.io.Serial
import javax.swing.Action
import javax.swing.JLabel

class RemoteHacConnectionDialog(
    private val project: Project,
    parentComponent: Component,
    private val settings: HybrisRemoteConnectionSettings
) : DialogWrapper(project, parentComponent, false, IdeModalityType.IDE) {

    private val originalScope = settings.scope
    private val testConnectionButton: Action = object : DialogWrapperAction("Test Connection") {

        @Serial
        private val serialVersionUID: Long = 7851071514284300449L

        override fun doAction(e: ActionEvent?) {
            val testSettings = with(HybrisRemoteConnectionSettings()) {
                type = settings.type
                hostIP = hostTextField.text
                port = portTextField.text
                isSsl = sslProtocolCheckBox.isSelected
                sslProtocol = sslProtocolComboBox.selectedItem?.toString() ?: ""
                hacWebroot = webrootTextField.text
                hacLogin = usernameTextField.text
                hacPassword = String(passwordTextField.password)
                this
            }

            val httpClient = HybrisHacHttpClient.getInstance(project)
            val errorMessage = httpClient.login(project, testSettings)

            val type: NotificationType
            val message: String
            if (errorMessage.isEmpty()) {
                message = message("hybris.toolwindow.hac.test.connection.success", "hac", testSettings.generatedURL)
                type = NotificationType.INFORMATION
            } else {
                type = NotificationType.WARNING
                message = message("hybris.toolwindow.hac.test.connection.fail", testSettings.generatedURL, errorMessage)
            }

            Notifications.create(type, message("hybris.notification.toolwindow.hac.test.connection.title"), message)
                .notify(project)
        }
    }

    private lateinit var connectionNameTextField: JBTextField
    private lateinit var hostTextField: JBTextField
    private lateinit var portTextField: JBTextField
    private lateinit var sslProtocolCheckBox: JBCheckBox
    private lateinit var sslProtocolComboBox: ComboBox<String>
    private lateinit var webrootTextField: JBTextField
    private lateinit var usernameTextField: JBTextField
    private lateinit var passwordTextField: JBPasswordField
    private lateinit var urlPreviewLabel: JLabel
    private val panel = panel {
        row {
            label("Connection name:")
                .bold()
            connectionNameTextField = textField()
                .align(AlignX.FILL)
                .bindText(settings::displayName.toNonNullableProperty(""))
                .component
        }.layout(RowLayout.PARENT_GRID)

        row {
            label("Scope:")
                .comment("Non-personal settings will be stored in the <strong>hybrisProjectSettings.xml</strong> and can be shared via VCS.")
            comboBox(
                EnumComboBoxModel(RemoteConnectionScope::class.java),
                renderer = SimpleListCellRenderer.create("?") { it.title }
            )
                .bindItem(settings::scope.toNullableProperty(RemoteConnectionScope.PROJECT_PERSONAL))
        }.layout(RowLayout.PARENT_GRID)

        group("Full URL Preview", false) {
            row {
                urlPreviewLabel = label(settings.generatedURL)
                    .bold()
                    .align(AlignX.FILL)
                    .component
            }
        }

        group("Host Settings") {
            row {
                label("Address:")
                    .comment("Host name or IP address")
                hostTextField = textField()
                    .align(AlignX.FILL)
                    .bindText(settings::hostIP.toNonNullableProperty(HybrisConstants.DEFAULT_HOST_URL))
                    .onChanged { urlPreviewLabel.text = generateUrl() }
                    .addValidationRule("Address cannot be blank.") { it.text.isNullOrBlank() }
                    .component
            }.layout(RowLayout.PARENT_GRID)

            row {
                label("Port:")
                portTextField = textField()
                    .align(AlignX.FILL)
                    .bindText(settings::port.toNonNullableProperty(""))
                    .onChanged { urlPreviewLabel.text = generateUrl() }
                    .addValidationRule("Port should be blank or in a range of 1..65535.") {
                        if (it.text.isNullOrBlank()) return@addValidationRule false

                        val intValue = it.text.toIntOrNull() ?: return@addValidationRule true
                        return@addValidationRule intValue !in 1..65535
                    }
                    .component
            }.layout(RowLayout.PARENT_GRID)

            row {
                sslProtocolCheckBox = checkBox("SSL:")
                    .selected(settings.isSsl)
                    .onChanged { urlPreviewLabel.text = generateUrl() }
                    .component
                sslProtocolComboBox = comboBox(
                    listOf(
                        "TLSv1",
                        "TLSv1.1",
                        "TLSv1.2"
                    ),
                    renderer = SimpleListCellRenderer.create("?") { it }
                )
                    .enabledIf(sslProtocolCheckBox.selected)
                    .bindItem(settings::sslProtocol.toNullableProperty())
                    .align(AlignX.FILL)
                    .component
            }.layout(RowLayout.PARENT_GRID)

            row {
                label("Webroot:")
                webrootTextField = textField()
                    .align(AlignX.FILL)
                    .bindText(settings::hacWebroot.toNonNullableProperty(""))
                    .onChanged { urlPreviewLabel.text = generateUrl() }
                    .component
            }.layout(RowLayout.PARENT_GRID)
        }

        group("Credentials") {
            row {
                label("Username:")
                usernameTextField = textField()
                    .align(AlignX.FILL)
                    .bindText(settings::hacLogin.toNonNullableProperty("admin"))
                    .addValidationRule("Username cannot be blank.") { it.text.isNullOrBlank() }
                    .component
            }.layout(RowLayout.PARENT_GRID)

            row {
                label("Password:")
                passwordTextField = passwordField()
                    .align(AlignX.FILL)
                    .enabled(false)
                    .applyToComponent {
                        ProgressManager.getInstance().run(object : Task.Backgroundable(project, "Retrieving password", false) {
                            private var password: String? = null
                            override fun run(indicator: ProgressIndicator) {
                                password = settings.hacPassword
                                passwordTextField.text = password
                                passwordTextField.isEnabled = true
                            }
                        })
                    }
                    .onApply {
                        ProgressManager.getInstance().run(object : Task.Backgroundable(project, "Persisting password", false) {
                            override fun run(indicator: ProgressIndicator) {
                                settings.hacPassword = String(passwordTextField.password)
                            }
                        })
                    }
                    .addValidationRule("Password cannot be blank.") { it.password.isEmpty() }
                    .component
            }.layout(RowLayout.PARENT_GRID)
        }
    }

    init {
        title = "Remote SAP Commerce Instance"
        super.init()
    }

    override fun applyFields() {
        super.applyFields()

        // change of the scope
        if (settings.uuid != null && originalScope != settings.scope) {
            RemoteConnectionUtil.changeRemoteConnectionScope(project, settings, originalScope)
        }
    }

    override fun createCenterPanel() = with(panel) {
        border = JBUI.Borders.empty(16)
        this
    }

    override fun createLeftSideActions() = arrayOf(testConnectionButton)
    override fun getStyle() = DialogStyle.COMPACT
    override fun getPreferredFocusedComponent() = connectionNameTextField

    private fun generateUrl() = RemoteConnectionUtil.generateUrl(
        sslProtocolCheckBox.isSelected,
        hostTextField.text,
        portTextField.text,
        webrootTextField.text
    )

}